Feature save and get AnalyticsEvent

DeemaP opened the pull request:
add '/api/v1' context path
add Interval enum
add base functions for AnalyticsEvent.
add Junit tests

DeemaP committed:
Feature save and get AnalyticsEvent

add '/api/v1' context path
add Interval enum
add base functions for AnalyticsEvent.
add Junit test

YTXIRE commented on src/main/java/faang/school/analytics/dto/analyticsEvent/AnalyticsEventFilterDto.java:
```diff
@@ -0,0 +1,15 @@
+import java.time.LocalDateTime;
+
+@Data
+@Builder
```
AllArgsConstructor не нужен тебе тут?)

DeemaP replied:
Добавил)

YTXIRE commented on src/main/java/faang/school/analytics/dto/analyticsEvent/AnalyticsEventResponseDto.java:
```diff
@@ -0,0 +1,17 @@
+
+@Data
+@Builder
+public class AnalyticsEventResponseDto {
```
И тут так же AllArgsConstructor)

DeemaP replied:
И тут)

YTXIRE commented on src/main/java/faang/school/analytics/filter/AnalyticsEventFromToFilter.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.stream.Stream;
+
+@Component
+public class AnalyticsEventFromToFilter implements Filter<AnalyticsEvent, AnalyticsEventFilterDto> {
```
Как мне кажется, что лучше разделить два фильтра, бывают случаи, когда нужно фильтрануть по дате от или по дате до.

DeemaP replied:
Дык я как раз так и сделал)
По условию задачи надо, что если Interval не задали, то применялись фильтры From To
И здесь они применяются по очереди. 
Если хочешь отфильтровать только from - пожалуйста, не пиши to)

YTXIRE commented on src/main/java/faang/school/analytics/model/Interval.java:
```diff
@@ -0,0 +1,37 @@
+import java.time.temporal.TemporalAmount;
+
+@Getter
+public enum Interval {
```
Почему данный класс в model лежит?
Модели - это же представление базы данных

DeemaP replied:
Я поглядел, у нас по всем микросервисам  enums лежат в entity или model (в зависимости от названия в конкретном сервисе)

Я думаю мы обсудим этот вопрос завтра на созвоне или может Руслан поправит, пока что я оставлю как есть)

RuslanZinovyev replied:
Можете создать пакет `enums `в папке модели  и положить туда, либо сделать вложенным классом другого класса если этим энумератором пользуется один конкретный класс

YTXIRE commented on src/main/java/faang/school/analytics/model/Interval.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public LocalDateTime calculateIntervalStart() {
+        LocalDateTime now = LocalDateTime.now(ZoneId.of("UTC"));
```
UTC+0? Может мск?

DeemaP replied:
Я везде у нас по проекту делал UTC+0
Независимо от того кто и где запустит проект) А то у нас команда от -8 до +7 :D

denis-levkhov commented on src/main/java/faang/school/analytics/dto/analyticsEvent/AnalyticsEventFilterDto.java:
```diff
@@ -0,0 +1,15 @@
+public class AnalyticsEventFilterDto {
+    Interval interval;
+    LocalDateTime from;
+    LocalDateTime to;
```
Может быть такое, что to будет раньше, чем from?

DeemaP replied:
прекрасное замечание, ща сделаю валидацию на это)

denis-levkhov commented on src/main/java/faang/school/analytics/filter/AnalyticsEventFromToFilter.java:
```diff
@@ -0,0 +1,32 @@
+        boolean isIntervalNull = filters.getInterval() == null;
+        boolean isFromOrToPresent = filters.getFrom() != null || filters.getTo() != null;
+
+        return isIntervalNull && isFromOrToPresent;
```
```java
return filters.getInterval() == null && (filters.getFrom() != null || filters.getTo() != null);
```
Почему бы не написать так?

DeemaP replied:
Я вообще выпилил фильтры))

DeemaP replied:
Но то, как я написал сделано для читаемости))

denis-levkhov commented on src/main/java/faang/school/analytics/model/Interval.java:
```diff
@@ -0,0 +1,37 @@
+    public LocalDateTime calculateIntervalStart() {
+        LocalDateTime now = LocalDateTime.now(ZoneId.of("UTC"));
+
+        if (interval instanceof Duration) {
```
Почему бы нельзя сделать так?
```java
public LocalDateTime calculateIntervalStart() {
    LocalDateTime now = LocalDateTime.now(ZoneId.of("UTC"));
    return now.minus(interval);
}
```

DeemaP replied:
Ох. я тут вчера закопался, действительно так можно сделать, ведь interval объявлен как TemporalAmount
а Duration и Period оба реализуют TemporalAmount, собсна все кайф)

Поправил

denis-levkhov commented on src/main/java/faang/school/analytics/service/AnalyticsEventService.java:
```diff
@@ -0,0 +1,57 @@
+        log.info("Analytics event {} saved successfully.", savedEvent.getId());
+    }
+
+    @Transactional(readOnly = true)
```
readOnly - круто, интересная штученция, изучу завтра утром

DeemaP replied:
Если честно, я ее указал, поскольку у меня приложение не запускалось, там из базы возвращается стрим, поэтому спринг требовал повесить эту аннотацию

В новой версии вообще нет @Transactional, поскольку вся фильтрация на стороне базы и в сервисе просто один запрос отправляется)

DeemaP committed:
Feature save and get AnalyticsEvent

Add validation for invalid date range in filtering
Remove unused filters
Switch AnalyticsEvent repository from CrudRepository to JpaRepository
Add Specification for filtering queries at the database level
Refactor AnalyticsEventService

DeemaP committed:
refactor

delete filterDto
delete @Transactional in service

RuslanZinovyev commented on src/main/java/faang/school/analytics/controller/AnalyticsEventController.java:
```diff
@@ -0,0 +1,46 @@
+public class AnalyticsEventController {
+    private final AnalyticsEventService analyticsEventService;
+
+    @GetMapping("/receiverId/{receiverId}/eventType/{eventType}")
```
`/receivers/{receiverId}/event-type/{eventType}`

DeemaP replied:
ой да, опять в ночи забыл про kebab-case

RuslanZinovyev commented on src/main/java/faang/school/analytics/controller/AnalyticsEventController.java:
```diff
@@ -0,0 +1,46 @@
+public class AnalyticsEventController {
+    private final AnalyticsEventService analyticsEventService;
+
+    @GetMapping("/receiverId/{receiverId}/eventType/{eventType}")
```
`/receivers/{receiverId}/event-type/{eventType}`

RuslanZinovyev commented on src/main/java/faang/school/analytics/service/AnalyticsEventService.java:
```diff
@@ -0,0 +1,57 @@
+            specification = specification.and(AnalyticsEventSpecification.withinPeriod(from, to));
+        }
+
+        List<AnalyticsEventResponseDto> result = analyticsEventRepository.findAll(specification).stream()
```
Должно быть производительнее если мы будем сортировать на уровне базы:

```
    Sort sort = Sort.by(Sort.Direction.DESC, "receivedAt");

    List<AnalyticsEventResponseDto> result = analyticsEventRepository.findAll(specification, sort).stream()
            .map(analyticsEventMapper::entityToResponseDto)
            .toList();
```

DeemaP replied:
Спасибо за подсказку! Реализовал

RuslanZinovyev commented on src/main/java/faang/school/analytics/specification/AnalyticsEventSpecification.java:
```diff
@@ -0,0 +1,34 @@
+            } else if (end != null) {
+                return criteriaBuilder.lessThanOrEqualTo(root.get("receivedAt"), end);
+            }
+            return null;
```
А можно сделать так чтобы null не возвращать?

```
public static Specification<AnalyticsEvent> withinPeriod(LocalDateTime start, LocalDateTime end) {
    return (root, query, criteriaBuilder) -> {
        Predicate predicate = criteriaBuilder.conjunction();

        if (start != null && end != null) {
            predicate = criteriaBuilder.between(root.get("receivedAt"), start, end);
        } else if (start != null) {
            predicate = criteriaBuilder.greaterThanOrEqualTo(root.get("receivedAt"), start);
        } else if (end != null) {
            predicate = criteriaBuilder.lessThanOrEqualTo(root.get("receivedAt"), end);
        }

        return predicate;
    };
}
```

DeemaP replied:
Спасибо за это уточнение!
Действительно так гораздо правильнее, ведь мы не бросаем нигде null, а возвращаем просто предикат с true внутри! Круто))

DeemaP committed:
refactor:
 implement changes based on reviewer feedback
